### PR TITLE
Add Glossary section

### LINKED ISSUE #1966: Add glossary definitions for coroutine terms
There are a few cases in the documentation and a few more in docstrings (and a lot of people I talk to) where the terms regarding coroutines are a bit muddled together. It would be nice to add a glossary to the documentation and explicitly define all the terms regarding coroutines.

* **coroutine function**: the definition of a function that when called returns a coroutine object. Implemented using async functions.
* **coroutine**: the result of calling an coroutine function. Coroutines are not run immediately, you must either: `await` on them which blocks the awaiting coroutine until it is finished; or fork the coroutine, turning it into a task, which runs concurrently.
* **task**: the result of forking a coroutine, it represents a concurrently running coroutine.

The old syntax uses `coroutine`s for coroutine functions, and calling them resulted in a `RunningCoroutine`, which confuses the matter.

### PR DESCRIPTION
**Pull request by cmarqu**
Closes #1966.

### CODE REVIEW

**Comment by cmarqu** [comment]
@ktbarrett Could you suggest something for "async function"?

File: documentation/source/glossary.rst [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+========
+
+.. glossary::
+
+   async function
+      FIXME
+
+   coroutine function
+      The definition of a function that, when called, returns a coroutine object.
+      Implemented using :term:`async functions<async function>`.
```
Comment (ktbarrett) [inline documentation/source/glossary.rst:13]: I would link back to Python's documentation on async syntax. Here is Python's glossary on [coroutines and coroutine functions](https://docs.python.org/3/glossary.html#term-coroutine) and [async def functions](https://docs.python.org/3/reference/compound_stmts.html#async-def).

File: documentation/source/glossary.rst [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+   coroutine
+      The result of calling a :term:`coroutine function`.
+      Coroutines are not run immediately, you must either:
+      await on them which blocks the awaiting coroutine until it is finished;
```
Comment (ktbarrett) [inline documentation/source/glossary.rst:18]: I seem to remember being able to crossref syntax, I may be wrong...

**Comment by cmarqu** [comment]
Looking good?: https://cocotb--1968.org.readthedocs.build/en/1968/glossary.html

File: documentation/source/glossary.rst [hunk 3]
```diff
@@ -0,0 +1,64 @@
+.. _glossary:
+
+Glossary
+========
+
+.. glossary::
+
+   coroutine function
+      The definition of a function that, when called, returns a coroutine object.
+      Implemented using :ref:`async functions<python:async def>`.
```
Comment (ktbarrett) [inline documentation/source/glossary.rst:10]:
```suggestion
      Implemented using :keyword:`async` functions.
      See also the :term:`Python glossary <python:coroutine function>`.
```

File: documentation/source/glossary.rst [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+   coroutine
+      The result of calling a :term:`coroutine function`.
+      Coroutines are not run immediately, you must either
+      :keyword:`await` on them which blocks the awaiting coroutine until it is finished;
+      or fork the coroutine, turning it into a :term:`task`, which runs concurrently.
+      See also the :term:`Python coroutine glossary <python:coroutine>`.
```
Comment (ktbarrett) [inline documentation/source/glossary.rst:17]:
```suggestion
      See also the :term:`Python glossary <python:coroutine>`.
```

**Pull request merged**